### PR TITLE
TELCODOCS-336 - firmware settings updates via GitOps ZTP

### PR DESCRIPTION
**Pull request by aireilly**
Adds docs for doing BIOS firmware settings updates via GitOps ZTP

Version(s):
4.17

Issue:
https://issues.redhat.com/browse/TELCODOCS-336

Link to docs preview:
https://81426--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/ztp-deploying-far-edge-sites.html#ztp-configuring-host-firmware-with-gitops-ztp_ztp-deploying-far-edge-sites

QE review:
- [ ] QE has approved this change.

### CODE REVIEW

File: modules/ztp-retrieving-the-host-firmware-settings.adoc [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    schema:
+      name: schema-40562318
+      namespace: compute-1
+    settings: <3>
+      AccessControlService: Enabled
+      AcpiHpet: Enabled
+      AcpiRootBridgePxm: Enabled
+      # ...
+----
+<1> Indicates that the host firmware has changed
```
Comment (tliu2021) [inline modules/ztp-retrieving-the-host-firmware-settings.adoc:78]: To make it more clear: This status indicates that a change in the host firmware settings has been detected.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+----
++
+.node-firmware.profile
+[source,yaml]
+----
+EnergyEfficientTurbo: Disabled
+ProcTurboMode: Enabled
+ProcPwrPerf: OsDbpm
+ProcC1E: Disabled
+ProcCStates: Enabled
```
Comment (tliu2021) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:55]: These profile examples from the QE Git repository are designed to test the hierarchy of profiles, where the node-level profile overrides the cluster-level profile, and the cluster-level profile overrides the site-level profile.
For the custom documentation, we only need one profile example. This example should demonstrate how to apply the profile at the site level, cluster level, or node level, while noting that the node-level profile will overwrite both the cluster-level and site-level profiles.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+[source,terminal]
+----
+example-ztp/install
+    └── site-install
+          ├── siteconfig-example.yaml
+          ├── kustomization.yaml
+          └── firmware-profiles
+                ├── site-firmware.profile
+                ├── cluster-firmware.profile
+                └── node-firmware.profile
```
Comment (tliu2021) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:69]: Same here, we only need to show one profile,

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+apiVersion: ran.openshift.io/v1
+kind: SiteConfig
+metadata:
+  name: "site-plan-cluster"
+  namespace: "example-cluster-namespace"
+spec:
+  baseDomain: "example.com"
+  # ...
+  biosConfigRef:
+    filePath: "./firmware-profiles/site-firmware.profile" <1>
```
Comment (tliu2021) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:86]: We add a note to say this is how to include the profile at the site level.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  clusters:
+    - clusterName: "cluster-1"
+      # ...
+      biosConfigRef:
+        filePath: "./firmware-profiles/cluster-firmware.profile" <2>
```
Comment (tliu2021) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:91]: Same as above, but for the cluster level.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      # ...
+      nodes:
+        - hostName: "compute-1.example.com"
+          # ...
+          bootMode: "UEFI"
+          biosConfigRef:
+            filePath: "./firmware-profiles/node-firmware.profile" <3>
```
Comment (tliu2021) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:98]: Same as above, but for site level.

**Comment by aireilly** [comment]
@imiller0 hi! Can you also PTAL at this.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc (hunk 3 above)
Comment (aireilly) [reply modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:69]: Ah ok I was confused about that actually 😂

File: modules/ztp-retrieving-the-host-firmware-settings.adoc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+.Procedure
+
+. Retrieve the host firmware settings for the managed cluster.
+Run the following command:
++
+--
+[source,terminal]
+----
+$ oc get hostfirmwaresettings -n <managed_cluster_namespace> -o yaml
```
Comment (imiller0) [inline modules/ztp-retrieving-the-host-firmware-settings.adoc:28]: Might be helpful to spec this as `oc get hostfirmwaresettings -n <cluster_namespace> <node_name> -o yaml` to demonstrate how to get it for just one node. In the case of SNO it's the same, but for SNO+1 or any multinode cluster you may want to be more specific.

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 8]
```diff
@@ -0,0 +1,129 @@
+// Module included in the following assemblies:
+//
+// * edge_computing/ztp-deploying-far-edge-sites.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp_{context}"]
+= Deploying user-defined firmware to cluster hosts with {ztp}
+
+You can deploy user-defined firmware to cluster hosts by configuring the `SiteConfig` custom resource (CR) to include a hardware profile that you want to apply during cluster host provisioning.
```
Comment (imiller0) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:9]: We should be really clear throughout that we are updating the firmware settings only and not doing a firmware update.
```suggestion
You can deploy user-defined firmware settings to cluster hosts by configuring the `SiteConfig` custom resource (CR) to include a hardware profile that you want to apply during cluster host provisioning.
```

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      # ...
+      nodes:
+        - hostName: "compute-1.example.com"
+          # ...
+          bootMode: "UEFI"
+          biosConfigRef:
+            filePath: "./firmware-profiles/node-firmware.profile" <3>
+# ...
+----
+<1> Applies to all hosts deployed site-wide
```
Comment (imiller0) [inline modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:101]: In the future we will support only one cluster per "SiteConfig". Maybe consider adding a note here like "(it is recommended to have only one cluster per SiteConfig CR)"

File: modules/ztp-configuring-host-firmware-with-gitops-ztp.adoc [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+Hosts require the correct firmware configuration to ensure high performance and optimal efficiency.
+You can deploy custom host firmware configurations for managed clusters with {ztp}.
+
+Tune hosts with specific hardware profiles in your lab and ensure they are optimized for your requirements.
+When you have completed host tuning to your satisfaction, you extract the host profile and save it in your {ztp} repository.
+Then, you use the host profile to patch and update managed cluster hosts that you deploy with {ztp}.
+
+You specify the required hardware profiles in `SiteConfig` custom resources (CRs) that you use to deploy the managed clusters.
+The {ztp} pipeline generates the required `HostFirmwareSettings` (`HFS`) and `BareMetalHost` (`BMH`) CRs that are applied to the hub cluster.
+The Bare Metal Operator that is running in the hub cluster uses the `HFS` controller to continuously reconcile `HFS` resources to ensure they match the expected state until the host is detached.
```
[the remark this answers is not in the document]
Comment (aireilly) [reply modules/ztp-configuring-host-firmware-with-gitops-ztp.adoc:18]: I think we can just remove this line

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc (hunk 8 above)
Comment (aireilly) [reply modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:9]: So we can change the settings, and the system itself does the actual firmware update itself upon reboot?

File: modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc (hunk 9 above)
Comment (aireilly) [reply modules/ztp-deploying-user-defined-firmware-configuration-with-gitops-ztp.adoc:101]: Seems kind of strange that we are documenting doing site-wide configurations but suggesting to not do this...

I've added a note but not sure it works tbh.

**Comment by dgonyier** [comment]
QE 
/lgtm

File: modules/ztp-configuring-host-firmware-with-gitops-ztp.adoc [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+Tune hosts with specific hardware profiles in your lab and ensure they are optimized for your requirements.
+When you have completed host tuning to your satisfaction, you extract the host profile and save it in your {ztp} repository.
+Then, you use the host profile to configure firmware settings in the managed cluster hosts that you deploy with {ztp}.
+
+You specify the required hardware profiles in `SiteConfig` custom resources (CRs) that you use to deploy the managed clusters.
+The {ztp} pipeline generates the required `HostFirmwareSettings` (`HFS`) and `BareMetalHost` (`BMH`) CRs that are applied to the hub cluster.
+
+Use the following best practices to manage your host firmware profiles.
+
+Identify critical firmware settings with hardware vendors::
```
Comment (kcarmichael08) [inline modules/ztp-configuring-host-firmware-with-gitops-ztp.adoc:21]: This is a nice use of formatting to highlight these. Gonna steal that!

**Review by kcarmichael08 — COMMENTED** [review]
Hey, I marked a few things as suggestions, but this looks great!

**Pull request merged**